### PR TITLE
Add nonhydro overflow

### PR DESCRIPTION
**Pull request by scalandr**
This PR consists of adding two test cases in the overflow test group. Their purpose is to run an overflow scenario using the nonhhydrostatic capability and show the difference between the hydrostatic and nonhydrostatic result. The `nonhydro` test builds a new mesh and initial conditions and runs the nonhydrostatic model for 30min. The `hydro_vs_nonhydro` test builds a new mesh and initial conditions and runs an hydrostatic and a nonhydrostatic simulation. A plot comparing the two temperature profiles after 3h is produced.
At this stage, the mesh and initial conditions for these two new tests are different (but similar) to those used in the default and rpe tests (already present in the overflow test group). @cbegeman and I talked about the possibility of using the mesh and initial conditions of the nonhydro tests also for the default and rpe tests after this review.

### CODE REVIEW

**Comment by scalandr** [comment]
At the moment, since the nonhydro PR (https://github.com/E3SM-Project/E3SM/pull/5706) is still in progress, please use this branch
https://github.com/scalandr/E3SM/tree/ocean/nonhydro
for testing.

**Comment by scalandr** [comment]
I ran both test cases on Chicoma and Anvil with the gnu compiler and the expected results are produced
section_overflow

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+        for j in range(nGrids):
+            grid = grids[j]
+            ncfileIC = Dataset(f'init_{grid}.nc', 'r')
+            ncfile = Dataset(f'output_{grid}.nc', 'r')
+            temp = ncfile.variables['temperature']
+            plt.subplot(2, 1, j + 1)
+            plt.imshow(temp[time, 0:nx, :].T)
+            plt.clim([10, 20])
+            plt.jet()
```
Comment (xylar) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py:57]: Could we use `cmo.thermal` instead

**Comment by xylar** [comment]
@scalandr, this looks fantastic, but I won't be able to do a thorough review for awhile.

@cbegeman, is this one you have time to review sooner? If so, no need to wait for my review.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        nVertices = ds.nVertices.size
+
+        xCell = ds.xCell
+
+        # comment('create and initialize variables')
+        time1 = time.time()
+
+        # bottom depth
+        # ds['bottomDepth'] = maxDepth * xarray.ones_like(xCell)
+        ds['bottomDepth'] = - (- 200.0 + 0.5 * (200.0 - 40.0) *
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:83]:
```suggestion
        ds['bottomDepth'] = - (- bottom_depth + 0.5 * (bottom_depth - min_depth) *
```
Is this right? If so, add `min_depth` to the cfg file and get both `min_depth` and `bottom_depth` from the cfg file.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                               (1.0 + np.tanh((6400.0 -
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:84]:
```suggestion
                               (1.0 + np.tanh((nx * dc -
```
Is this right?

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        # ssh
+        ds['ssh'] = xarray.zeros_like(xCell)
+
+        init_vertical_coord(config, ds)
+
+        # initial salinity, density, temperature
+        ds['salinity'] = (config_eos_linear_Sref *
+                          xarray.ones_like(ds.zMid)).where(ds.cellMask)
+        ds['density'] = (config_eos_linear_densityref +
+                         rhoz * ds.zMid).where(ds.cellMask)
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:95]:
```suggestion
```
Let's just choose initial salinity and temperature profiles that give you the `rhoz` you want and remove `rhoz` from the cfg file. I suspect that this density field isn't actually used by MPAS-Ocean but rather that density is recomputed from temperature and salinity even at the first time step.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        # T = Tref - (rho - rhoRef)/alpha
+        ds['temperature'] = xarray.ones_like(ds.zMid).where(ds.cellMask)
+        temperature = ds['temperature']
+        for iCell in range(0, nCells):
+            temperature[0, iCell, :] = -1.0
+            for k in range(0, nVertLevels):
+                if (xCell[iCell] < 990.0):
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:102]: Can 990.0 be expressed in terms of the cfg options? That way, if someone decides to change the cfg options the colder temperature ends up in the right place.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                    temperature[0, iCell, k] = lower_temperature
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:103]: Hmm, does temperature actually have 3 dimensions? I would have thought that `zMid` was missing the time dimension.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        print(f'   time: {time.time() - time1}')
+
+        # comment('finalize and write file')
+        time1 = time.time()
+
+        # If you prefer not to have NaN as the fill value, you should consider
+        # using mpas_tools.io.write_netcdf() instead
+        write_netcdf(ds, 'initial_state.nc')
+        print(f'   time: {time.time() - time1}')
+        print(f'Total time: {time.time() - timeStart}')
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:128]: When you're done testing, let's delete the timer-related lines.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/namelist.forward [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+config_dt = '00:00:0.5'
+config_time_integrator = 'RK4'
+config_use_mom_del2 = .true.
+config_mom_del2 = 1.0
+config_use_mom_del4 = .false.
+config_vert_tracer_adv_flux_order = 3
+config_horiz_tracer_adv_order = 2
+config_coef_3rd_order = 0.25
+config_flux_limiter = 'monotonic'
+config_remap_limiter = 'monotonic'
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/namelist.forward:12]:
```suggestion
```
You don't need this line if you're not using vertical Lagrangian-remapping

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/namelist.forward [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+config_enable_nonhydrostatic_mode = .false.
+config_use_cvmix = .true.
+config_disable_vel_coriolis = .false.
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/namelist.forward:15]:
```suggestion
```
I think the default is `.false.` so you shouldn't need this line

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        Create the step
+
+        Parameters
+        ----------
+        test_case : compass.TestCase
+            The test case this step belongs to
+        """
+        super().__init__(test_case=test_case, name='visualize')
+
+        for grid in ['nonhydro', 'hydro']:
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py:27]:
```suggestion
        for mode in ['nonhydro', 'hydro']:
```
I think it's clearer to change `grid` to `mode` here and below

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            grid = grids[j]
+            ncfileIC = Dataset(f'init_{grid}.nc', 'r')
+            ncfile = Dataset(f'output_{grid}.nc', 'r')
+            temp = ncfile.variables['temperature']
+            plt.subplot(2, 1, j + 1)
+            plt.imshow(temp[time, 0:nx, :].T)
+            plt.clim([10, 20])
+            plt.jet()
+            plt.colorbar()
+            plt.xticks([0, 50, 100, 150, 200, 250, 300], [0, 1, 2, 3, 4, 5, 6])
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py:59]:
```suggestion
            plt.xticks(np.arange(0., nx, 1000. / dc), np.arange(0, nx * dc / 1000., 1))
```
Double-check that I've written this correctly. I'm thinking about something like this here and below so that the plot comes out ok if someone changes the domain size.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        grids = ['hydro', 'nonhydro']
+        nGrids = len(grids)
+        plt.figure(1, figsize=(12.0, 6.0))
+
+        config = self.config
+
+        section = config['horizontal_grid']
+        nx = section.getint('nx')
+        section = config['visualize']
+        time = section.getint('plotTime')
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py:47]:
```suggestion
        time = section.getint('plot_time')
```
Also, see earlier comment about writing `plot_time` in hours

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/hydro_vs_nonhydro.cfg [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+#Lower temperature (deg C)
+lower_temperature = 10.0
+
+#Higher temperature (deg C)
+higher_temperature = 20.0
+
+# config options for the solitary wave testcase
+[visualize]
+
+#Time at which plotting the variables
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/hydro_vs_nonhydro.cfg:61]:
```suggestion
# Time index at which to plot the variables
```
This might be more convenient if written in hours so that the user doesn't have to check the streams file to see how frequently variables are output.

File: compass/ocean/tests/overflow/nonhydro/initial_state.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
+import time
```
Comment (cbegeman) [inline compass/ocean/tests/overflow/nonhydro/initial_state.py:1]: If this file is the same as `nonhydro_vs_hydro/initial_state`, let's just retain one of those files and import it for use in the other test case. You might even move it up one directory and rename my `initial_state` step to `initial_state_from_init_mode` or something like that.

**Review by cbegeman — CHANGES_REQUESTED** [review]
@scalandr Thank you so much for working on this! It will be awesome to replace the init mode step with something similar to what you have written. Overall, it looks great! I made some suggestions to make the code more readable and maintainable. Let me know what you think. I haven't done any testing yet; I'll plan to do that after you make the suggested changes.

**Comment by cbegeman** [comment]
> @scalandr, this looks fantastic, but I won't be able to do a thorough review for awhile.
> 
> @cbegeman, is this one you have time to review sooner? If so, no need to wait for my review.

Yep! Just reviewed.

File: compass/ocean/tests/overflow/__init__.py [hunk 15]
```diff
@@ -17,6 +19,8 @@ def __init__(self, mpas_core):
 
         self.add_test_case(Default(test_group=self, resolution='10km'))
         self.add_test_case(RpeTest(test_group=self, resolution='2km'))
+        self.add_test_case(Nonhydro(test_group=self))
+        self.add_test_case(HydroVSNonhydro(test_group=self))
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/overflow/__init__.py:23]: Nope, I changed it as suggested.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/hydro_vs_nonhydro.cfg [hunk 16]
```diff
@@ -0,0 +1,62 @@
+# config options for the horizontal grid
+[horizontal_grid]
+
+#Number of cells in the x-direction
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/hydro_vs_nonhydro.cfg:4]: Done.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/hydro_vs_nonhydro.cfg [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+lower_temperature = 10.0
+
+#Higher temperature (deg C)
+higher_temperature = 20.0
+
+# config options for the solitary wave testcase
+[visualize]
+
+#Time at which plotting the variables
+plotTime = 6
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/hydro_vs_nonhydro.cfg:62]: Change made.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+        nEdges = ds.nEdges.size
+        nVertices = ds.nVertices.size
+
+        xCell = ds.xCell
+
+        # comment('create and initialize variables')
+        time1 = time.time()
+
+        # bottom depth
+        # ds['bottomDepth'] = maxDepth * xarray.ones_like(xCell)
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:82]: Yes, line deleted.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+        # vertical_coordinate = 'uniform'
+
+        ds = dsMesh.copy()
+        nCells = ds.nCells.size
+        nEdges = ds.nEdges.size
+        nVertices = ds.nVertices.size
+
+        xCell = ds.xCell
+
+        # comment('create and initialize variables')
```
[the remark this answers is not in the document]
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:78]: Done.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py (hunk 2 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:83]: `bottom_depth` and `shelf_depth` are now read from the cfg file.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/namelist.forward (hunk 8 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/namelist.forward]: Done.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/namelist.forward (hunk 9 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/namelist.forward]: Done.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py (hunk 12 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py]: Change made.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py (hunk 7 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:130]: Done.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py (hunk 5 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:104]: Added among the cfg options.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py (hunk 4 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:97]: Deleted.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py (hunk 3 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:86]: Yes, I changed it.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/hydro_vs_nonhydro.cfg (hunk 13 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/hydro_vs_nonhydro.cfg]: Modified comment.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py (hunk 10 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py:27]: Change made.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py (hunk 11 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py:59]: Line changed.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py (hunk 6 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/initial_state.py:95]: zMid does have the time dimension.

File: compass/ocean/tests/overflow/nonhydro/initial_state.py (hunk 14 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/nonhydro/initial_state.py]: I moved `initial_state` up one directory and renamed yours `initial_state_from_init_mode` as you suggested.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py (hunk 1 above)
Comment (cbegeman) [inline compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py:59]:
```suggestion
            plt.imshow(temp[time, 0:nx, :].T, cmap='cmo.thermal', vmin=10., vmax=20.)
```
This is all you would need to do, in addition to adding `import cmocean`

**Review by cbegeman — APPROVED** [review]
This test case is looking good to me. There's just the final viz fix that @xylar suggested. I ran the two new cases and created the docs and everything looks as expected.

In the near future, I think we'll also want to make a single forward step that works for all the cases. I'm fine leaving this for the next PR where we hopefully will be getting rid of init mode altogether.

**Comment by xylar** [comment]
@scalandr, as I said on Slack, I'm happy to merge this as soon as we agree on resolving the color map.  `cmo.thermal` is the standard colormap for temperature in oceanography so we prefer to use it unless there's a solid reason not to.  We recommend not using linear rather maps (like `cmo.thermal`) rather than diverging maps like `jet` for quantities where the central value in the colorbar doesn't have a particular physical meaning, as is the case for temperature.  A divergent colormap is the right thing for a signed quantity like a velocity component.

Another reason to avoid `jet` and other "rainbow" maps is that people who are colorblind find the colors hard to distinguish from what I understand.  This is why both Matlab and Python (`matplotlib`) have switched to a linear colormap like `viridis` as their default.

File: compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py (hunk 1 above)
Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py:59]: Hi @cbegeman, I changed the `imshow` line and added `import cmocean`, but when I try to commit it tells me
`compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py:1:1: F401 'cmocean' imported but unused`
should I just get rid of `import cmocean`?

Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py:59]: This is how the new plot looks like:
section_overflow

Comment (scalandr) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py]: Done, thank you @xylar! Should I request a re-review?

Comment (xylar) [reply compass/ocean/tests/overflow/hydro_vs_nonhydro/visualize.py]: No need.  Thanks!

**Review by xylar — APPROVED** [review]
@scalandr, wonderful work.  What an exciting new capability!

**Comment by cbegeman** [comment]
Congrats, @scalandr!
